linker/Linker.Steps/MarkStep.cs

```diff
@@ -1895,10 +2002,30 @@ protected virtual void HandleUnresolvedMethod (MethodReference reference)
 			}
 		}
 
-		protected virtual void MarkInterfaceImplementation (TypeDefinition type, InterfaceImplementation iface)
+		protected virtual bool ShouldMarkInterfaceImplementation (TypeDefinition type, InterfaceImplementation iface, TypeDefinition resolvedInterfaceType)
+		{
+			if (Annotations.IsMarked (iface))
+				return false;
+
+			if (Annotations.IsMarked (resolvedInterfaceType) && !Annotations.IsMarked (iface))
+				return true;
+
+			// It's hard to know if a com or windows runtime interface will be needed from managed code alone,
+			// so as a precaution we will mark these interfaces once the type is instantiated
+			if (resolvedInterfaceType.IsImport || resolvedInterfaceType.IsWindowsRuntime)
+				return true;
+
+			if (!Annotations.IsPreserved (type))
```

Comment:
Looks like a redundant check

Reply:
The next method call, `Annotations.GetPreserve (type)` throws if `!Annotations.IsPreserved (type)`.   I'd be more comfortable with the call to `Annotations.GetPreserve` guarded by the 

```
if (!Annotations.IsPreserved (type))
    return false;
```

I'm not 100% confident this is 100% redundant.  Maybe I'm just missing it.  I would rather err on the side of caution than depend on implicit conditions based on other state we track.

Do you want to elaborate on why you think it's redundant?

The fact that `Annotations.GetPreserve` throws is a constant annoyance to me.  I understand why it does, but a lot of the times I find myself querying this information I do not want anything to throw.  Maybe the solution is to add a `TryGetPreserve` method?

Reply:
Let's clean this in separate PR by renaming GetPreserve to TryGetPreserve. The uses of `GetPreserve` are not always correct either.